### PR TITLE
app: initial work to get building on Windows

### PR DESCRIPTION
**Pull request by limitedmage**
App now builds on Windows with the command `enterprise/dev/app/build-windows.sh`. To set it up, see `doc/dev/background-information/app/windows.md`. It still doesn't run (issues with embedded Postgresql).

Note that the build works, but the bundling into an installer does not, because we use year-based version numbers for our app, and MSI needs version numbers to be less than 255.

## Test plan

App should build on Windows with `enterprise/dev/app/build-windows.sh` after installing required dependencies.

### CODE REVIEW

File: cmd/searcher/internal/search/mmap_windows.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+	mmapgo "github.com/edsrzf/mmap-go"
+)
+
+func mmap(path string, f *os.File, fi fs.FileInfo) ([]byte, error) {
+	return mmapgo.Map(f, mmapgo.RDONLY, 0)
+}
+
+func unmap(data mmapgo.MMap) error {
+	return data.Unmap()
+}
```
[the remark this answers is not in the document]
Comment (limitedmage) [reply cmd/searcher/internal/search/mmap_windows.go:16]: Yeah, this confused me too. Turns out `mmapgo.MMap` is a type alias for `[]byte` so it does work.

[the remark this answers is not in the document]
Comment (camdencheek) [reply cmd/searcher/internal/search/mmap_windows.go:16]: ah, got it!

**Comment by camdencheek** [comment]
@limitedmage is this in an "I think this is mergeable" state now? If so, I'm happy to give it a real review

**Comment by limitedmage** [comment]
Yes, I think this is OK to merge as long as it doesn’t break anything. the
regular Linux/Mac build. We can keep iterating on this after merging.

File: cmd/gitserver/server/cleanup.go [hunk 2]
```diff
@@ -20,6 +20,7 @@ import (
 
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/ricochet2200/go-disk-usage/du"
```
Comment (camdencheek) [inline cmd/gitserver/server/cleanup.go:23]: This is...not the most trustworthy looking package 😅 

Have you tested it? In particular, docker can do weird things with disk reporting.

File: enterprise/cmd/executor/internal/run/setup_ip_tables_windows.go [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package run
+
+import "github.com/sourcegraph/sourcegraph/enterprise/cmd/executor/internal/util"
+
+func SetupIPTables(runner util.CmdRunner, recreateChain bool) error {
+	return nil
```
Comment (camdencheek) [inline enterprise/cmd/executor/internal/run/setup_ip_tables_windows.go:6]: Rather than returning `nil`, I think it would be better to `panic` here. My understanding is this should never actually be called, and if it is, then we're probably doing something wrong and we should be loud about it.

File: enterprise/cmd/executor/internal/worker/workspace/unmount_windows.go [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package workspace
+
+import "errors"
+
+func unmount(dirPath string) error {
+	return errors.New("unmount not supported on Windows")
```
Comment (camdencheek) [inline enterprise/cmd/executor/internal/worker/workspace/unmount_windows.go:6]: similarly, a panic here is probably appropriate since an error can be ignored more easily

File: go.mod [hunk 5]
```diff
@@ -254,18 +254,20 @@ require (
...
 	layeh.com/gopher-luar v1.0.10
 	sigs.k8s.io/kustomize/kyaml v0.13.3
 	sigs.k8s.io/yaml v1.3.0
 )
 
 require (
+	github.com/edsrzf/mmap-go v1.1.0
 	github.com/go-redsync/redsync/v4 v4.8.1
 	github.com/jackc/pgerrcode v0.0.0-20220416144525-469b46aa5efa
+	github.com/ricochet2200/go-disk-usage/du v0.0.0-20210707232629-ac9918953285
```
Comment (camdencheek) [inline go.mod:270]: Any reason to not pin a [released version](https://github.com/ricochet2200/go-disk-usage/releases)?

File: internal/singleprogram/postgresql.go [hunk 6]
```diff
@@ -112,7 +113,7 @@ func startEmbeddedPostgreSQL(logger log.Logger, pgRootDir string) (*postgresqlEn
 			RuntimePath(filepath.Join(pgRootDir, "runtime")).
 			Username(vars.PGUSER).
 			Database(vars.PGDATABASE).
-			UseUnixSocket(unixSocketDir).
+			UseUnixSocket(strings.ReplaceAll(unixSocketDir, "\\", "/")).
```
Comment (camdencheek) [inline internal/singleprogram/postgresql.go:116]: Why is this needed?

**Review by camdencheek — APPROVED** [review]
So did things like compiling sqlite just...work?

**Comment by limitedmage** [comment]
> So did things like compiling sqlite just...work?

I mean, this compiles, but not sure how much of it is working yet, since Postgresql isn't connecting yet 🫠

File: cmd/gitserver/server/cleanup.go (hunk 2 above)
Comment (limitedmage) [reply cmd/gitserver/server/cleanup.go:23]: I agree on the trustworthiness 😅 but I looked at the code and it's incredibly lightweight for non-Windows, doing the same things we were doing before: https://github.com/ricochet2200/go-disk-usage/blob/master/du/diskusage.go

File: internal/singleprogram/postgresql.go (hunk 6 above)
Comment (limitedmage) [reply internal/singleprogram/postgresql.go:116]: This is me trying to (unsuccessfully) get Postgresql to work :( I'll revert it.

File: cmd/gitserver/server/cleanup.go (hunk 2 above)
Comment (camdencheek) [reply cmd/gitserver/server/cleanup.go:23]: Cool, sounds good 👍

File: go.mod (hunk 5 above)
Comment (limitedmage) [reply go.mod:270]: Because I don't know how to go :P

Comment (camdencheek) [reply go.mod:270]: Try this:

```suggestion
	github.com/ricochet2200/go-disk-usage/du v1.1.2
```

Comment (camdencheek) [reply go.mod:270]: (that can also be done with `go get github.com/ricochet2200/go-disk-usage/du@v1.1.2`

Comment (camdencheek) [reply go.mod:270]: oh, actually, this is not a you problem, it's a problem with the dependency

Comment (camdencheek) [reply go.mod:270]: They have a separate go.mod file for the `./du` directory, which makes it a separate module with a separate versioning, so there is no way to version the `du` package nicely 🤦

Comment (limitedmage) [reply go.mod:270]: Ugh :(

Comment (camdencheek) [reply go.mod:270]: It's not a huge deal. What you have will work fine, just a weird style 👍
